libsupport/include/katana/ArrowVisitor.h

```diff
@@ -0,0 +1,333 @@
+  using ParamType = const typename arrow::TypeTraits<ArrowType>::ScalarType&;
+
+  template <arrow::Type::type EnumType>
+  static const auto& Cast(const arrow::Scalar& scalar) {
+    using ArrowType = typename arrow::TypeIdTraits<EnumType>::Type;
+    return static_cast<ParamType<ArrowType>>(scalar);
+  }
+
+  static std::shared_ptr<arrow::DataType> Type(const arrow::Scalar& scalar) {
+    return scalar.type;
+  }
+};
+
+// Builders are mutable, pass by pointer
+struct BuilderVisitorBaseType {
+  using ParamBase = arrow::ArrayBuilder*;
+
+  template <typename ArrowType>
+  using ParamType = typename arrow::TypeTraits<ArrowType>::BuilderType*;
+
+  template <arrow::Type::type EnumType>
+  static auto Cast(arrow::ArrayBuilder* builder) {
+    using ArrowType = typename arrow::TypeIdTraits<EnumType>::Type;
+    return static_cast<ParamType<ArrowType>>(builder);
+  }
+
+  static std::shared_ptr<arrow::DataType> Type(arrow::ArrayBuilder* builder) {
+    return builder->type();
+  }
+};
+
+/// This function attempts to keep Arrow type-tests to a minimum by
+/// encapsulating the desired behavior. Its behavior is to:
+/// 1) Identify the Arrow type of the parameter
+/// 2) Specialize the parameter for use in a template environment
+/// 3) Call the templated visitor with the specialized parameter
+///
+/// In an ideal world, this is the *only* switch over Arrow types
+/// in the whole repo. Everything should either handle arrow data
+/// in a type-agnostic way, or use this interface to reach a
+/// type-aware template environment. It probably recreates
+/// functionality that exists elsewhere, notably in arrow::compute
+///
+/// A typical use case would involve creating a Visitor class
+/// and implementing methods to handle various types, eg:
+/// class Visitor {
+///   using ReturnType = void; // configurable
+///
+///   template <typename ArrowType, typename ArgumentType>
+///   katana::Result<ReturnType> Call(ArgumentType arg);
+/// };
+/// Arrow's type_traits.h offers tools to use SFINAE to differentiate
+/// types and write appropriate Call functions for each type
+template <class VisitorBaseType, class VisitorType>
+katana::Result<typename std::decay_t<VisitorType>::ReturnType>
+VisitArrowInternal(
+    typename VisitorBaseType::ParamBase param, VisitorType&& visitor) {
+  switch (VisitorBaseType::Type(param)->id()) {
+#define TYPE_CASE(EnumType)                                                    \
+  case arrow::Type::EnumType: {                                                \
+    using ArrowType =                                                          \
+        typename arrow::TypeIdTraits<arrow::Type::EnumType>::Type;             \
+    return visitor.template Call<ArrowType>(                                   \
+        VisitorBaseType::template Cast<arrow::Type::EnumType>(param));         \
+  }
+    TYPE_CASE(UINT8)
+    TYPE_CASE(INT32)
+    TYPE_CASE(UINT32)
+    TYPE_CASE(INT64)
+    TYPE_CASE(UINT64)
+    TYPE_CASE(FLOAT)
```

Comment:
Hey.  I had to add `UINT8 `to support `LDBC_003 `and given that this list is the list of supported updatable types for customer properties, is there any problem with me adding the full complement of `U?Int{8,16,32,64}` types here?

Reply:
Should be fine, they'll all fall under `is_number`